sdk/typescript/src/bcs/index.js

```diff
@@ -0,0 +1,663 @@
+            value: total,
+            length: len
+        };
+    }
+
+    // define constants 
+    MoveBCS.types = new Map();
+    MoveBCS.U8 = 'u8';
+    MoveBCS.U32 = 'u32';
+    MoveBCS.U64 = 'u64';
+    MoveBCS.U128 = 'u128';
+    MoveBCS.BOOL = 'bool';
+    MoveBCS.VECTOR = 'vector';
+    MoveBCS.ADDRESS = 'address';
+    MoveBCS.STRING = 'string';
+
+    MoveBCS.hasType = function hasType(type) {
+        return MoveBCS.types.has(type);
+    };
+
+    /**
+     * Method to register new types for BCS internal representation.
+     * For each registered type 2 callbacks must be specified and one is optional:
+     * 
+     * - encodeCb(writer, data) - write a way to serialize data with BcsWriter;
+     * - decodeCb(reader) - write a way to deserialize data with BcsReader;
+     * - validateCb(data) - validate data - either return bool or throw an error
+     * 
+     * @example
+     * // our type would be a string that consists only of numbers
+     * BCS.registerType('number_string',
+     *    (writer, data) => writer.writeVec(data, (w, el) => w.write8(el)),
+     *    (reader) => reader.readVec((r) => r.read8()).join(''), // read each value as u8
+     *    (value) => /[0-9]+/.test(value) // test that it has at least one digit
+     * );
+     * console.assert(BCS.de('number_string', BCS.ser('number_string', '12345').hex()) === '12345');
+     * 
+     * @param {String} name
+     * @param {Function} encodeCb 
+     * @param {Function} decodeCb Callback used to 
+     * @param {?Function} validateCb Optional validator Callback to check type before serialization
+     */
+    MoveBCS.registerType = function registerType(
+        name, 
+        encodeCb, 
+        decodeCb, 
+        validateCb = () => true
+    ) {
+        this.types.set(name, {
+            encode(data, size = 1024) { return this._encodeRaw(new MoveBCS.BcsWriter(size), data); },
+            decode(data) { return this._decodeRaw(new MoveBCS.BcsReader(data)); },
+
+            // these methods should always be used with caution as they require pre-defined
+            // reader and writer and mainly exist to allow multi-field (de)serialization;
+            _encodeRaw(writer, data) {
+                if (validateCb(data)) {
+                    return encodeCb(writer, data)
+                } else {
+                    throw new Error(`Validation failed for type ${name}, data: ${data}`);
+                }
+            },
+            _decodeRaw(reader) { return decodeCb(reader); }
+        });
+    };
+
+    /**
+     * Register custom vector type inside the MoveBCS. 
+     * 
+     * @example
+     * MoveBCS.registerVectorType('vector<u8>', 'u8');
+     * let array = MoveBCS.de('vector<u8>', MoveBCS.util.fromHex('06010203040506')); // [1,2,3,4,5,6];
+     * let again = MoveBCS.ser('vector<u8>', [1,2,3,4,5,6]).toBytes();
+     * 
+     * @param {String} name Name of the type to register.
+     * @param {String} innerType Name of the inner type of the vector.
```

Comment:
nit: could we name this `elementType` ?  

with collections in strongly typed languages that support generics, you can easily get type defs like `Vec<Option<u64>>`.  
In this case, what would be the "inner type" of `Vec<Option<u64>>` - `Option<u64>` or `u64` ?

the `element` label I believe more clearly indicates exactly which type is being referred to - the one immediately contained by the `Vec`

Reply:
It is the next logical step, but I want to first settle on the core functionality.